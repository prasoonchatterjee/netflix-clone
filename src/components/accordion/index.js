import React, { useState, useContext, createContext } from "react";
import {
  Container,
  Title,
  Frame,
  Item,
  Inner,
  Header,
  Body,
} from "./styles/accordion";
const ToggleContext = createContext();

export default function Accordion({ children, ...restProps }) {
  return (
    <Container {...restProps}>
      <Inner>{children}</Inner>
    </Container>
  );
}

Accordion.Title = function AccordionTitle({ children, ...restProps }) {
  return <Title {...restProps}>{children}</Title>;
};

Accordion.Frame = function AccordionFrame({ children, ...restProps }) {
  return <Frame {...restProps}>{children}</Frame>;
};

Accordion.Item = function AccordionItem({ children, ...restProps }) {
  const [toggleShow, setToggeleShow] = useState(false);
  return (
    <ToggleContext.Provider value={{ toggleShow, setToggeleShow }}>
      <Item {...restProps}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...restProps }) {
  const { toggleShow, setToggeleShow } = useContext(ToggleContext);
  return (
    <Header
      onClick={() => setToggeleShow((toggleShow) => !toggleShow)}
      {...restProps}
    >
      {children}
      {toggleShow ? (
        <img src="/images/icons/close-slim.png" alt="Close" />
      ) : (
        <img src="/images/icons/add.png" alt="Open" />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...restProps }) {
  const { toggleShow } = useContext(ToggleContext);

  return toggleShow && <Body {...restProps}>{children}</Body>;
};
