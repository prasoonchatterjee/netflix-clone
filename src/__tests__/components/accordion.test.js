import React from "react";
import { render, fireEvent } from "@testing-library/react";
import faqsData from "../../fixtures/faqs.json";
import { Accordion } from "../../components";

describe("<Accordion />", () => {
  test("renders <Accordion /> with populated data", () => {
    const { container, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqsData.map((item) => (
          <Accordion.Item key={item.id}>
            <Accordion.Header>{item.header}</Accordion.Header>
            <Accordion.Body>{item.body}</Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>
    );

    expect(getByText("Frequently Asked Questions")).toBeTruthy();
    expect(getByText("What is Netflix?")).toBeTruthy();
    expect(getByText("How much does Netflix cost?")).toBeTruthy();
    expect(getByText("Where can I watch?")).toBeTruthy();
    expect(getByText("How do I cancel?")).toBeTruthy();
    expect(getByText("What can I watch on Netflix?")).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("opens and closes the <Accordion/> component", () => {
    const { container, queryByTestId, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        {faqsData.map((item) => (
          <Accordion.Item key={item.id}>
            <Accordion.Header>{item.header}</Accordion.Header>
            <Accordion.Body data-testid="accordion body">
              {item.body}
            </Accordion.Body>
          </Accordion.Item>
        ))}
      </Accordion>
    );

    expect(queryByTestId("accordion body")).toBeFalsy();
    fireEvent.click(getByText("What is Netflix?"));
    expect(queryByTestId("accordion body")).toBeTruthy();
    fireEvent.click(getByText("What is Netflix?"));
    expect(queryByTestId("accordion body")).toBeFalsy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
