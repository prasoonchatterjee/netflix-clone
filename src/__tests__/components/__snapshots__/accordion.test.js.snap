// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> opens and closes the <Accordion/> component 1`] = `
<div
  class="accordion__Container-sc-1wnyt6p-0 cNKPUD"
>
  <div
    class="accordion__Inner-sc-1wnyt6p-2 ggIgBn"
  >
    <h1
      class="accordion__Title-sc-1wnyt6p-3 ddnMFY"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        What is Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        How much does Netflix cost?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        Where can I watch?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        How do I cancel?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        What can I watch on Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Accordion /> renders <Accordion /> with populated data 1`] = `
<div
  class="accordion__Container-sc-1wnyt6p-0 cNKPUD"
>
  <div
    class="accordion__Inner-sc-1wnyt6p-2 ggIgBn"
  >
    <h1
      class="accordion__Title-sc-1wnyt6p-3 ddnMFY"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        What is Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        How much does Netflix cost?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        Where can I watch?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        How do I cancel?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
    <div
      class="accordion__Item-sc-1wnyt6p-4 gxtzye"
    >
      <div
        class="accordion__Header-sc-1wnyt6p-5 iuJlae"
      >
        What can I watch on Netflix?
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;
